src/actions/actions.ts

```diff
@@ -590,10 +590,17 @@ class CommandInsertInInsertMode extends BaseCommand {
         vimState.cursorStartPosition = leftPosition;
       }
     } else {
-      await TextEditor.insert(char, vimState.cursorPosition);
+      const selection = vscode.window.activeTextEditor.selection;
+      if (selection.isEmpty) {
+        await TextEditor.insert(char, vimState.cursorPosition);
+      } else {
+        await TextEditor.replace(selection, char);
+      }
+
+      const cursorPosition = vscode.window.activeTextEditor.selection.end;
 
-      vimState.cursorStartPosition = Position.FromVSCodePosition(vscode.window.activeTextEditor.selection.start);
-      vimState.cursorPosition = Position.FromVSCodePosition(vscode.window.activeTextEditor.selection.start);
+      vimState.cursorStartPosition = Position.FromVSCodePosition(cursorPosition);
```

Comment:
> Actually: Is there a case where we don't want the .cursorPosition (line below) be .selection.end?

Yeah, cursorPosition is generally equal to selection.end. However, we usually don't do that because that would require waiting for VSCode to update the position, which can take a long time. The speed doesn't generally matter that much except for things like . and macros, in which case it becomes very noticeable.